rpcserver.go

```diff
@@ -3185,6 +3189,91 @@ func (r *rpcServer) SubscribeReceiveAssetEventNtfns(
 	)
 }
 
+// SubscribeReceiveEvents registers a subscription to the event notification
+// stream which relates to the asset receiving process.
+func (r *rpcServer) SubscribeReceiveEvents(
+	req *taprpc.SubscribeReceiveEventsRequest,
+	ntfnStream receiveEventStream) error {
+
+	tapParams := address.ParamsForChain(r.cfg.ChainParams.Name)
+
+	// We just decode the address to make sure it's valid. But any
+	// comparison for filtering happens on the string representation, as
+	// that's easily comparable
+	var addrString string
+	if req.FilterAddr != "" {
+		addr, err := address.DecodeAddress(req.FilterAddr, &tapParams)
+		if err != nil {
+			return fmt.Errorf("error decoding address: %w", err)
+		}
+
+		addrString, err = addr.EncodeAddress()
+		if err != nil {
+			return fmt.Errorf("error encoding address: %w", err)
+		}
+	}
+
+	marshaler := func(event fn.Event) (*taprpc.ReceiveEvent, error) {
+		e, ok := event.(*tapgarden.AssetReceiveEvent)
+		if !ok {
+			return nil, fmt.Errorf("invalid event type: %T", event)
+		}
+
+		rpcAddr, err := marshalAddr(&e.Address, r.cfg.TapAddrBook)
+		if err != nil {
+			return nil, fmt.Errorf("error marshaling addr: %w", err)
+		}
+
+		rpcStatus, err := marshalAddrEventStatus(e.Status)
+		if err != nil {
+			return nil, fmt.Errorf("error marshaling status: %w",
+				err)
+		}
+
+		return &taprpc.ReceiveEvent{
+			Timestamp: e.Timestamp().UnixMicro(),
+			Address:   rpcAddr,
+			Outpoint:  e.OutPoint.String(),
+			Status:    rpcStatus,
+		}, nil
+	}
+
+	filter := func(event fn.Event) (bool, error) {
+		var (
+			eventAddrString string
+			err             error
+		)
+
+		switch e := event.(type) {
+		case *tapgarden.AssetReceiveEvent:
+			eventAddrString, err = e.Address.EncodeAddress()
+			if err != nil {
+				return false, fmt.Errorf("error encoding "+
+					"address: %w", err)
+			}
+
+		case *proof.BackoffWaitEvent:
+			// We're not interested in any backoff events.
+			return false, nil
+
+		default:
+			return false, fmt.Errorf("unknown event type: %T", e)
+		}
+
+		// If we're not filtering on a specific address, we return all
+		// events.
+		if addrString == "" {
+			return true, nil
+		}
+
+		return eventAddrString == addrString, nil
+	}
+
+	return handleEvents[bool, *taprpc.ReceiveEvent](
```

Comment:
Could also consider a another refactor where we make a new interface for the filter+marshal methods, so then the event handler accepts that struct/impl instead of these function closures. 

Non-blocking.

Reply:
Do you mean a method on the actual event object (for example `tapgarden.AssetReceiveEvent`) itself so it knows how to serialize itself to RPC?
Or a new struct in the RPC server, something like `receiveEventFilterMarshaler`?
I'll consider that for part, but I feel like things are already pretty compact and readable the way they are now.

Reply:
So basically morph `handleEvents` into a a struct that can accept the generic params of filter type, marshaller, etc.